qiskit/circuit/__init__.py

```diff
@@ -54,81 +78,825 @@
+:class:`.expr.Var` node.
+
+A circuit can contain manual classical storage locations, represented internally by the
+:class:`~.expr.Var` node of the :class:`.Expr` tree.  These have an attached classical type (like
+any other expression).  These can either be declared and initialized within each execution of the
+circuit (:meth:`~QuantumCircuit.add_var`), or be inputs to the circuit
+(:meth:`~QuantumCircuit.add_input`).
+
+.. _circuit-compile-time-parameters:
+
+Compile-time parametrization
+----------------------------
+
+Various parametric :class:`Instruction` instances in Qiskit can be parametrized in ways that are
+designed to be resolved at compile time.  These are characterized by the use of the
+:class:`Parameter` and :class:`ParameterExpression` classes.
+
+.. autoclass:: Parameter
+.. autoclass:: ParameterExpression
+
+The main way that this differs from the :class:`expr.Var` variables used in near-time classical
+computation is that :class:`ParameterExpression` is a symbolic representation of a mathematical
+expression.  The semantics of the expression are those of regular mathematics over the continuous
+real numbers (and, in limited cases, over the complex numbers). In contrast, :class:`.Var` is a
+handle to a variable stored on a classical computer, such as a floating-point value or an
+fixed-width integer, which are always discrete.
+
+In other words, you can expect :class:`ParameterExpression` to do symbolic simplifications that are
+valid in mathematics, such as simplifying :math:`(x + y - x) / y \to 1`.  Such a simplification is
+not valid in floating-point arithmetic, and :class:`.expr.Expr` will not do this.
+
+The "compile-time" part of these parameters means that you typically will want to "assign" values to
+the parameters before sending them for execution.  These parameters can typically be used anywhere
+that expects a mathematical angle (like a rotation gate's parameters), with the caveat that hardware
+will usually require them to be assigned to a proper classically typed value before execution.  You
+can do this assignment using :meth:`QuantumCircuit.assign_parameters`.
+
+You may want to use many parameters that are related to each other.  To make this easier (and to
+avoid you needing to come up with many names), you can use the convenience constructor
+:class:`ParameterVector`.  The elements of the vector are all valid :class:`Parameter` instances.
+
+.. autoclass:: ParameterVector
+
+.. _circuit-control-flow:
+
+Control flow in circuits
+------------------------
+
+Within :class:`QuantumCircuit`, classical control flow is represented by specific
+:class:`Instruction`\ s, which are subclasses of :class:`ControlFlowOp`.
+
+.. autoclass:: ControlFlowOp
+    :members:
+    :no-inherited-members:
+
+These control-flow operations (:class:`IfElseOp`, :class:`WhileLoopOp`,
+:class:`SwitchCaseOp` and :class:`ForLoopOp`) all have specific state that defines the branching
```

Comment:
```suggestion
:class:`SwitchCaseOp` and :class:`ForLoopOp`) all have a specific state that defines the branching
```

Reply:
I'm using "state" in the sense of "class state" - classes "have state" in the sense of "are stateful" rather than "have a state".